fedora_messaging/cli.py

```diff
@@ -168,3 +174,176 @@ def consume(exchange, queue_name, routing_key, callback, app_name):
+            try:
+                api.publish(msg, exchange)
+            except exceptions.PublishReturned as e:
+                click.echo("Unable to publish message: {}".format(str(e.reason)))
+                sys.exit(errno.EREMOTEIO)
+            except exceptions.ConnectionException as e:
+                click.echo(
+                    "Unable to connect to the message broker: {}".format(str(e.reason))
+                )
+                sys.exit(errno.ECONNREFUSED)
+
+
+class Recorder:
+    """
+    A simple callback class that records messages.
+
+    Attributes:
+        counter (int): The number of messages this callback has recorded.
+        messages (list): The list of messages received.
+
+    Args:
+        limit (int): The maximum number of messages to record.
+        file (file): The file object with write rights.
+    """
+
+    def __init__(self, limit, file):
+        self.counter = 0
+        self.messages = []
+        self._limit = limit
+        self._file = file
+        if limit:
+            self._bar = click.progressbar(length=limit)
+
+        def _signal_handler(signum, frame):
+            """
+            Signal handler that gracefully terminates the recorder
+
+            Args:
+                signum (int): The signal this process received.
+                frame (frame): The current stack frame (unused).
+            """
+            if signum in (signal.SIGTERM, signal.SIGINT):
+                self.save_messages_to_file()
+                raise exceptions.HaltConsumer(exit_code=0, requeue=False)
+
+        signal.signal(signal.SIGTERM, _signal_handler)
+        signal.signal(signal.SIGINT, _signal_handler)
+
+    def collect_message(self, message):
+        """
+        Collect received messages.
+
+        Args:
+            message (message.Message): The received message.
+
+        Raises:
+            fedora_messaging.exceptions.HaltConsumer: Raised if the number of received
+                messages reach the limit.
+        """
+        self.messages.append(message)
+        self.counter += 1
+        if self._limit:
+            self._bar.update(1)
+            if self._limit <= self.counter:
+                self.save_messages_to_file()
+                raise exceptions.HaltConsumer(exit_code=0, requeue=False)
+
+    def save_messages_to_file(self):
+        """
+        Save all collected messages to file.
+
+        Raises:
+            fedora_messaging.exceptions.HaltConsumer: Raised if collected messages serialization
+                is impossible.
+        """
+        # Reset the signal handler
+        for signum in (signal.SIGTERM, signal.SIGINT):
+            signal.signal(signum, signal.SIG_DFL)
+        try:
+            json_str = dumps(self.messages)
+        except TypeError as e:
+            click.echo("Unable to save messages to file: {}".format(str(e)))
+            raise exceptions.HaltConsumer(exit_code=1, requeue=False)
```

Comment:
If we can't write the message to the file, maybe we want the message to be requeued here?

Reply:
If dumping raises a TypeError I believe it's not valid JSON so dropping it seems appropriate to me. Although it seems like this code would never be executed in practice since the consume loop would drop the message when json-schema failed on it, so it would never be passed to the callback.

Reply:
Ah right, then the error message above is misleading.